### PR TITLE
Replace if valid operator for initial temperature conditions

### PR DESCRIPTION
**Pull request by SophieCoulson**
Create new 'replace if valid' operator for initial temperature conditions. It checks if the value in the second model in a nan, if so it replaces with the value from the first model.

Test also included based on the tests for the other operators.

### CODE REVIEW

File: include/aspect/utilities.h [hunk 1]
```diff
@@ -1121,7 +1121,8 @@ namespace aspect
           add,
           subtract,
           minimum,
-          maximum
+          maximum,
+		  replace_if_valid
```
Comment (bangerth) [inline include/aspect/utilities.h]: You'll have to run the indent script I guess :-)

File: source/utilities.cc [hunk 2]
```diff
@@ -2794,6 +2794,12 @@ namespace aspect
           {
             return std::max(x,y);
           }
+          case Utilities::Operator::replace_if_valid:
+		  {
+        	  if (std::isnan(y))
+			  return x;
+			  else return y;
```
Comment (bangerth) [inline source/utilities.cc]: same here

File: source/utilities.cc [hunk 3]
```diff
@@ -2826,10 +2832,12 @@ namespace aspect
             operator_list[i] = Operator(Operator::minimum);
           else if (operator_names[i] == "maximum")
             operator_list[i] = Operator(Operator::maximum);
+          else if (operator_names[i] == "replace if valid")
+            operator_list[i] = Operator(Operator::replace_if_valid);
           else
             AssertThrow(false,
                         ExcMessage ("ASPECT only accepts the following operators: "
-                                    "add, subtract, minimum and maximum. But your parameter file "
+                                    "add, subtract, minimum, maximum and replace if valid. But your parameter file "
```
Comment (bangerth) [inline source/utilities.cc]:
```suggestion
                                    "add, subtract, minimum, maximum, and replace if valid. But your parameter file "
```

File: tests/initial_temperature_list_using_replace_if_valid.prm [hunk 4]
```diff
@@ -0,0 +1,86 @@
+##### simple test for using multiple initial temperature plugins
```
Comment (bangerth) [inline tests/initial_temperature_list_using_replace_if_valid.prm]: can you add a sentence that describes what's specific about this test?

**Comment by SophieCoulson** [comment]
Thanks Wolfgang - these things should be fixed now

**Comment by tjhei** [comment]
I have one comment, that we can do as a separate PR: The operator classs is used in several places in the code and those places need adjusting (to add the additional option in the ``prm.declare_entry`` call):
```
$ grep create_model_operator_list -r source/ include/
source/boundary_temperature/interface.cc:        model_operators = Utilities::create_model_operator_list(model_operator_names);
source/initial_composition/interface.cc:        model_operators = Utilities::create_model_operator_list(model_operator_names);
source/boundary_composition/interface.cc:        model_operators = Utilities::create_model_operator_list(model_operator_names);
source/initial_temperature/interface.cc:        model_operators = Utilities::create_model_operator_list(model_operator_names);

```
Ideally, we move the string into a helper function in Utilities (something like ``const std::string get_model_operator_options()`` that returns the string ""add|subtract|minimum|maximum|replace if valid". That way, they can not become out of sync. Again, this can be a followup as the code is correct right now.

File: tests/initial_temperature_list_using_replace_if_valid.prm [hunk 5]
```diff
@@ -0,0 +1,87 @@
...
+# This tests whether the 'replace if valid' operator is working
+'
```
Comment (tjhei) [inline tests/initial_temperature_list_using_replace_if_valid.prm]: remove this please, it causes the test to fail.

File: source/utilities.cc [hunk 6]
```diff
@@ -2794,6 +2794,12 @@ namespace aspect
           {
             return std::max(x,y);
           }
+          case Utilities::Operator::replace_if_valid:
+          {
+            if (std::isnan(y))
+              return x;
+            else return y;
```
Comment (bangerth) [inline source/utilities.cc]:
```suggestion
            else
              return y;
```

**Comment by SophieCoulson** [comment]
I've accepted Wolfgang's change and removed the header lines in the test file. Will do the other pull request in due course.

**Comment by tjhei** [comment]
The test fails with
``Message:  <Unexpected token "Nan" found at position 10.>``. Are you sure it works on your PC?

**Comment by SophieCoulson** [comment]
Changed Nan to lower case nan, the tester is running again

**Comment by tjhei** [comment]
So, on my machine (and the tester), we can not parse any of "NaN, nan, inf" and I am also unable to define these as additional constants, as you would need to write something like "NaN=NaN", which would require the thing on the right to be parseable. :-(
@bangerth can we discuss what to do at some point? I see the following options:
1. also allow something like -1e300 to register as invalid.
2. Ignore the fact that this new settings doesn't work with function parser. Modify the test to use a different plugin.
3. figure out a way to extend the function parser somehow. Maybe just search and replace "NaN" before parsing? :-)

**Comment by gassmoeller** [comment]
What was the result of the discussion about the NaNs?

File: source/initial_temperature/interface.cc [hunk 7]
```diff
@@ -204,7 +204,7 @@ namespace aspect
                           std::get<dim>(registered_plugins).get_description_string());
 
         prm.declare_entry("List of model operators", "add",
-                          Patterns::MultipleSelection("add|subtract|minimum|maximum"),
+                          Patterns::MultipleSelection("add|subtract|minimum|maximum|replace if valid"),
```
Comment (bangerth) [inline source/initial_temperature/interface.cc]: We should document "somewhere" what these operators actually are. Is there a good place for this?

**Comment by bangerth** [comment]
Ugh. Not sure what to do. Does it work on *some* systems at least?

I'd be ok to treat everything less than -1e300 as a nan. @SophieCoulson -- do you think you could make that addition (i.e., extend the test in the operator)? I think while you're there, you might also want to check whether the value is `inf` or `-inf`, since these are presumably also invalid values?

**Comment by SophieCoulson** [comment]
@bangerth @tjhei @gassmoeller 
Timo tested inf and -inf and neither of these work on microsoft either. I solved the problem by never specifying 'nan' in a parameter file (since I don't see why we'd ever need to do this in practice anyway). See PR #3007 
I'll close this pull request if you think that solution is okay.

**Pull request merged**